content/source/assets/javascripts/application.js

```diff
@@ -11,48 +11,132 @@
+            if ($("#docs-sidebar a.subnav-toggle").length === 0) { // ...in default location
+                $("#docs-sidebar #controls-placeholder").replaceWith(sidebarControlsHTML);
+            } else { // ...at a manually chosen location
+                $("#docs-sidebar a.subnav-toggle").replaceWith(sidebarControlsHTML);
+            }
         }
-        globalExpand = !globalExpand;
-    });
+
+        var subnavToggle = $("#sidebar-controls a.subnav-toggle");
+        var filterField = $("#sidebar-controls input#sidebar-filter-field");
+
+        // Expand/reset button behavior:
+        subnavToggle.on({
+            "taint": function(e) {
+                $(this).html("Reset");
+            },
+            "reset": function(e) {
+                filterField.val("");
+                filterField.trigger("blur");
+                sidebarLinks.parent("li").show();
+                resetActiveSubnavs();
+                $(this).html("Expand all");
+            },
+            "click": function(e) {
+                e.preventDefault();
+                if ($(this).text() === "Expand all") {
+                    subNavs.addClass("active");
+                    $(this).trigger("taint");
+                } else {
+                    $(this).trigger("reset");
+                }
+            }
+        });
+
+        // Filter as you type. This alters three things:
+        // - "active" class on subnavs
+        // - direct show/hide of <li>s
+        // - state of subnavToggle button
+        // We rely on subnavToggle's "reset" event to clean up when done.
+        filterField.on('keyup', function(e) {
+            if (e.keyCode === 27) { // escape key
+                subnavToggle.trigger("reset");
+            } else {
+                subnavToggle.trigger("taint");
+                var filterRegexp = new RegExp(filterField.val(), 'i');
+                var matchingLinks = sidebarLinks.filter(function(index) {
+                    return $(this).text().match(filterRegexp);
+                });
+                sidebarLinks.parent('li').hide();
+                subNavs.removeClass('active'); // cleans up partial as-you-type searches
+                // make matches and their parents visible and expanded:
+                matchingLinks.parents('li').show().filter(subNavs).addClass('active');
+                // make direct children visible (if your search caught a subnav directly):
+                matchingLinks.parent('li').find('li').show();
+            }
+        });
+        // Type slash to focus sidebar filter:
+        $(document).keydown(function(e) {
+            var focusedElementType = $(document.activeElement).get(0).tagName.toLowerCase();
+            var inputting = focusedElementType === "textarea" || focusedElementType === "input";
+
+            // / (forward slash) key = search
+            if (!inputting && e.keyCode === 191) {
+                e.preventDefault();
+                filterField.focus();
+            }
```

Comment:
So we will be allocating a few new variables every keystroke with this logic, it might be worth adding an earlier bailout, like `if (e.keyCode !== 191) return` at the top to filter out the majority of these situations?

Reply:
Oh also another thing that might be worth doing is scoping this only to pages that contain the sidebar. I assume you will get a javascript error on pages that don't (like the homepage) in its current state.

Reply:
Oh hey, bailing early's an excellent idea, thank you. 

I think this ought to be scoped only to sidebar pages — the whole thing's wrapped in an `if` statement, and the condition expands to `$("#docs-sidebar ul.nav.docs-sidenav").find('a')` not being empty. That's definitely empty on the homepage, and going through `ul.nav.docs-sidenav` means it's also empty on pages that use `inner.erb` as their layout but don't provide sidebar content (like /security.html).

Reply:
OH, nope, you're right. Turbolinks. \*shakes fist\* It's a handler on `document`, which outlives `<body>`.